### DISCUSSION TITLE
2.0.4 create_engine broken for sqlite?

### DISCUSSION THREAD

**Discussion by ViVe-IT** [sqlalchemy/sqlalchemy#9379] [Usage Questions]
Hi,

Just update to the latest avaialble version for Python 3.8

Works fine with 1.4.16 and completely breaks with 2.0.4:
``engine = sa.create_engine('sqlite://')``

with below error message:
``def set_regexp(dbapi_connection):        
dbapi_connection.create_function(
            "regexp", 2, regexp, **create_func_kw
        )
E       TypeError: function takes at most 3 arguments (4 given)``

Would be grateful for any hint on how to fix this

**Comment by CaselIT** [comment]
Hi,

What version of sqlite do you have in your environment?

**Reply by CaselIT** [reply]
Also you may have a python version that's not standard, since the docs specifies that create_function accept that signature:
https://docs.python.org/3/library/sqlite3.html?highlight=create_function#sqlite3.Connection.create_function

**Comment by zzzeek** [comment]
have never seen this before, if you could run this maybe we can guess

```
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.40.0'
>>> sqlite3.version
'2.6.0'
>>> 
```

**Reply by zzzeek** [reply]
also please provide complete stack trace

**Comment by ViVe-IT** [comment]
Hi,

Sqlite:
>>> import sqlite3
>>> sqlite3.version
'2.6.0'
>>> sqlite3.sqlite_version
'3.8.11'

The Python Version is the Standard RHEL Release of Python 3.8.12

the full stack trace:
```
dbapi_connection = <sqlite3.Connection object at 0x7fe6b1fd68f0>

    def set_regexp(dbapi_connection):
       dbapi_connection.create_function(
            "regexp", 2, regexp, **create_func_kw
        )
E       TypeError: function takes at most 3 arguments (4 given)
```
```
engine = sa.create_engine('sqlite://')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sqlalchemy/engine/base.py:3245: in connect
    return self._connection_cls(self)
sqlalchemy/engine/base.py:145: in __init__
    self._dbapi_connection = engine.raw_connection()
sqlalchemy/engine/base.py:3269: in raw_connection
    return self.pool.connect()
sqlalchemy/pool/impl.py:420: in connect
    return _ConnectionFairy._checkout(self, self._fairy)
sqlalchemy/pool/base.py:1270: in _checkout
    fairy = _ConnectionRecord.checkout(pool)
sqlalchemy/pool/base.py:719: in checkout
    rec = pool._do_get()
sqlalchemy/pool/impl.py:402: in _do_get
    c = self._create_connection()
sqlalchemy/pool/base.py:396: in _create_connection
    return _ConnectionRecord(self)
sqlalchemy/pool/base.py:681: in __init__
    self.__connect()
sqlalchemy/pool/base.py:917: in __connect
    pool.dispatch.connect.for_modify(
sqlalchemy/event/attr.py:473: in _exec_w_sync_on_first_run
    self(*args, **kw)
sqlalchemy/event/attr.py:487: in __call__
    fn(*args, **kw)
sqlalchemy/engine/create.py:718: in on_connect
    do_on_connect(dbapi_connection)
sqlalchemy/dialects/sqlite/pysqlite.py:567: in connect
    fn(conn)
```

**Reply by zzzeek** [reply]
> The Python Version is the Standard RHEL Release of Python 3.8.12

Nevertheless please share sqlite_version, I work at Red Hat and we certainly would have picked up a SQLAlchemy incompatibility such as this.

Here's 3.8.3 Red Hat:

```
Python 3.8.3 (default, May 23 2020, 16:34:37) 
[GCC 9.3.1 20200408 (Red Hat 9.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> e = create_engine("sqlite://")
>>> conn = e.connect()
>>> conn.exec_driver_sql("select 1").all()
[(1,)]

```

**Reply by zzzeek** [reply]
also can you please install SQLAlchemy cleanly in a fresh virtual environment, ensure no errant .pyc files are lying around

the error shows we are sending three positional arguments

**Reply by zzzeek** [reply]
The change in code is we are adding the "deterministic" parameter, which is available in Python 3.8:

https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.create_function

Here's main:

https://github.com/sqlalchemy/sqlalchemy/blob/1dadf3584ab9d3a34544b8cb40f787927e55c789/lib/sqlalchemy/dialects/sqlite/pysqlite.py#L547-L553

Here's 1.4:

https://github.com/sqlalchemy/sqlalchemy/blob/81dd922cf79f4df0eafceb2e5092915d79d72853/lib/sqlalchemy/dialects/sqlite/pysqlite.py#L556-L566

**Reply by ViVe-IT** [reply]
Sqlite info: 
sqlite3.version
'2.6.0'
sqlite3.sqlite_version
'3.8.11'

Interesting, started with a completely new venv like you suggested, the create_engine is not at fault, it seems to be the connect instead:

```
Python 3.8.12 (default, Mar 10 2022, 10:58:29)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44.0.3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> e = create_engine("sqlite://")
>>> conn = e.connect()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3245, in connect
    return self._connection_cls(self)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3269, in raw_connection
    return self.pool.connect()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 420, in connect
    return _ConnectionFairy._checkout(self, self._fairy)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 1270, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 719, in checkout
    rec = pool._do_get()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 402, in _do_get
    c = self._create_connection()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 396, in _create_connection
    return _ConnectionRecord(self)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 681, in __init__
    self.__connect()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 917, in __connect
    pool.dispatch.connect.for_modify(
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 718, in on_connect
    do_on_connect(dbapi_connection)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/pysqlite.py", line 567, in connect
    fn(conn)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/pysqlite.py", line 550, in set_regexp
    dbapi_connection.create_function(
TypeError: function takes at most 3 arguments (4 given)
```

**Reply by CaselIT** [reply]
You seem to be using a non standard python install.

Can you test the following?
```py
import sqlite3
conn = sqlite3.connect(':memory:')
conn.create_function('regexp', 2, lambda a,b: a, deterministic=True)
```

If that fail it suggest the issue is in your python library, so you should re-install it or report the issue to whoever maintains it

**Reply by ViVe-IT** [reply]
Leads to the exact same issue, unfortunately I have

```
Python 3.8.12 (default, Mar 10 2022, 10:58:29)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44.0.3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> conn = sqlite3.connect(':memory:')
>>> conn.create_function('regexp', 2, lambda a,b: a, deterministic=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: function takes at most 3 arguments (4 given)
>>>
```

But I tested it on another Server as well (fresh venv but slightly newer python 3.8.14 and this got me at least another error message:
```
Python 3.8.14 (default, Sep 19 2022, 09:47:48)
[GCC 9.1.1 20190605 (Red Hat 9.1.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> conn = sqlite3.connect(':memory:')
>>> conn.create_function('regexp', 2, lambda a,b: a, deterministic=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.NotSupportedError: deterministic=True requires SQLite 3.8.3 or higher
```
Still inconsistent though since my first Server runs a newer version of sqlite (3.8.11)

**Reply by CaselIT** [reply]
ok maybe you are hitting this case
> [NotSupportedError](https://docs.python.org/3/library/sqlite3.html?highlight=create_function#sqlite3.NotSupportedError) – If deterministic is used with SQLite versions older than 3.8.3.

do you have any chance of updating sqlite3?

@zzzeek we have best effor support from 3.7.16 of sqlite, but that not really true, since this is broken on py3.8 with sqlite 3.8.3. I'm fine changing that min version or fixing this (would prefer changing the min version)

**Reply by ViVe-IT** [reply]
Will try to do so (RHEL 7 is pretty much end of life anyway), if possible (unfortunately not my decision alone) , I will update the discussion if this fixed the TypeError, if not SQLAlchemy-1.4.46 it is for now. Thanks so far!!

**Reply by zzzeek** [reply]
What is not great is we could be simply catching this NotSupportedError but they aren't raising it consistently

**Reply by zzzeek** [reply]
we have to fix this, we can't raise the min version here, we support back to Python 3.7 still and if there are 3.8s shipping with an old sqlite like this (again like on centos) we really should be handling this.

**Reply by CaselIT** [reply]
ok, then we have to check both py38 and sqlite version in that condition

**Reply by zzzeek** [reply]
yep :|

**Comment by CaselIT** [comment]
@nurglexxx can you try this patch to see if it fixes your issue? https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4476

Thanks!

**Reply by ViVe-IT** [reply]
Hi, unfortunately the issue still persists (I updated it directly in the site package, that's correct?), though it points to the "set_regexp" function now, stacktrace:

`Python 3.8.12 (default, Mar 10 2022, 10:58:29)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44.0.3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> e = create_engine("sqlite://")
>>> conn = e.connect()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3245, in connect
    return self._connection_cls(self)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3269, in raw_connection
    return self.pool.connect()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 420, in connect
    return _ConnectionFairy._checkout(self, self._fairy)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 1270, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 719, in checkout
    rec = pool._do_get()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 402, in _do_get
    c = self._create_connection()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 396, in _create_connection
    return _ConnectionRecord(self)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 681, in __init__
    self.__connect()
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 917, in __connect
    pool.dispatch.connect.for_modify(
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 718, in on_connect
    do_on_connect(dbapi_connection)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/pysqlite.py", line 572, in connect
    fn(conn)
  File "/EDR_ICT/.virtualenvs/sqlalchemy_test/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/pysqlite.py", line 555, in set_regexp
    dbapi_connection.create_function(
TypeError: function takes at most 3 arguments (4 given)`

**Reply by CaselIT** [reply]
Either the python docs are wrong, indicating 3.8.3 as the min sqlite version or that install is not standard.

If you change the compare in the patch to use `(3, 9)` instead of `(3,8,3)` to read `if util.py38 and self._get_server_version_info(None) >= (3, 9):` does it work?

**Reply by ViVe-IT** [reply]
@CaselIT Works! 

"if util.py38 and self._get_server_version_info(None) > (3, 9):"

Python 3.8.12 (default, Mar 10 2022, 10:58:29)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44.0.3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> e = create_engine("sqlite://")
>>> conn = e.connect()
>>> conn.exec_driver_sql("select 1").all()
[(1,)]

**Reply by CaselIT** [reply]
ok, let's go with that then. thanks for testing!

Not sure who is at fault here (python docs or redhat python distribution), but since both are very old versions of sqlite it should not matter much to use a more conservative check

**Reply by ViVe-IT** [reply]
just fyi: I checked, this is not the python version from the rhel repo (my bad, the other server's I tested it with have the repo version) but I compiled it quite some timer ago on the machine itself (but with the download directly from the python.org website, which should still amount to "Standard")

Thanks again for all the help! (and the great library!)

**Reply by zzzeek** [reply]
> just fyi: I checked, this is not the python version from the rhel repo (my bad, the other server's I tested it with have the repo version) but I compiled it quite some timer ago on the machine itself (but with the download directly from the python.org website, which should still amount to "Standard")

oh.   That explains a lot.

@CaselIT this makes me a little more sympathetic to "dont support this old SQLite version".   Of course if someone built an ancient version of sqlite itself then built a newer python against it, they're going to have problems.  That also means that Python 3.8, 3.9, 3.11, whatever, doesn't matter.  

I'm concerned with sqlite's that are *shipping in packages* right now.    

I just looked at https://sqlite.org/chronology.html and SQLite 3.8.11 is from 2015.      though it's maybe a little late to bump "best effort" to 3.9 considering we released 2.0 already.

**Reply by CaselIT** [reply]
I don't think anyone is going to complain if we change that, since the matrix of old sqlite version - supported python release is very small.
But since it's not that big of a change we may as well support it, since it's preventing connection that's the definition of "best effort"

**Reply by ViVe-IT** [reply]
Just to be clear, the SQLite is the one it shipped for rhel 7 and the corresponding CentOS Installation (I tested with both Servers), just the one python I tested with was compiled from the python.org library
